### PR TITLE
Python3 compatibility

### PR DESCRIPTION
**Pull request by markyoder**
- used Python "2to3" to do most of the heavy lifting for Python2.x to 3.x. 
- several additional changes were necessary, including some changes to calls to the unit testing library
- modified the testing base class to support Python 2.x and 3.x
- all tests except test_gdpower() pass. the gdpower() test output is not totally ridiculous, so it is not clear if perhaps the sample to which the generated output is flawed.

### CODE REVIEW

**Comment by jaidevd** [comment]
@markyoder You've accidentally added a lot of *.bak files in the commit. Please remove them and add them to the .gitignore.

**Comment by markyoder** [comment]
Fair point. I'll sort it out in the AM.

Mark

On Wednesday, January 6, 2016, Jaidev Deshpande <notifications@github.com>
wrote:

> @markyoder <https://github.com/markyoder> You've accidentally added a lot
> of *.bak files in the commit. Please remove them and add them to the
> .gitignore.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-signal/pytftb/pull/82#issuecomment-169253939>.
>


-- 
Mark Yoder, PhD
805 451 8750
https://www.linkedin.com/pub/mark-yoder/11/855/300

"If you bore me, you lose your soul..."
~belly

**Comment by jaidevd** [comment]
Thanks @markyoder Can you please resolve the merge conflict? Also, I think it would be better to have this code in a separate branch for now, since even the Python 2 code so far needs a little bit of work before we can have a beta release.

**Comment by jaidevd** [comment]
And test_gdpower() failing was my fault. Should be fixed now, as of https://github.com/scikit-signal/pytftb/commit/85a9a38acde6bfd7d812082eb92179e29a6ac33d

No idea what made me think that the ideal output was what it was.

**Comment by markyoder** [comment]
Ok. I'll merge your changes back to my repo and go from there.

On Wednesday, January 13, 2016, Jaidev Deshpande <notifications@github.com>
wrote:

> And test_gdpower() failing was my fault. Should be fixed now, as of
> 85a9a38
> <https://github.com/scikit-signal/pytftb/commit/85a9a38acde6bfd7d812082eb92179e29a6ac33d>
>
> No idea what made me think that the ideal output was what it was.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-signal/pytftb/pull/82#issuecomment-171255998>.
>


-- 
Mark Yoder, PhD
805 451 8750
https://www.linkedin.com/pub/mark-yoder/11/855/300

"If you bore me, you lose your soul..."
~belly

**Comment by markyoder** [comment]
merge conflicts are resolved; new "expected values," as per jaidedevd, are included; my comments regarding the failure of the test script are removed. as per py2/py3 branches, i am operating on a single branch; i have just taken the plunge into a committed python3 environment. it seems to me establishing py2/py3 branches is best achieved on the upstream side. alternatively, reverse compatibility can be integrated, though from the nature of the code, my guess is that it should be universally compatible: print() is supported in py2/3; outstanding incompatibilities (which may or may not be present) are range() --> list(range()), xrange() --> range(), a mess of urllib_x --> urllib, etc.

File: tftb/tests/test_misc.py [hunk 1]
```diff
@@ -50,22 +58,54 @@ def test_mexhat(self):
...
+        # let's be a bit smarter about this:
+        #actuals = misc.gdpower(8, 0.5)
+        actuals = misc.gdpower(len(ideal_sig), 0.5)
+        #
         for i, ideal in enumerate(ideals):
             actual = actuals[i]
-            np.testing.assert_allclose(ideal, actual, atol=1e-7, rtol=1e-7)
+            #
+            #np.testing.assert_allclose(ideal, actual, atol=1e-7, rtol=1e-7)
+            if diagnostic:
```
Comment (jaidevd) [inline tftb/tests/test_misc.py]: @markyoder This seems to be some code added for debugging. Please remove it and all the plotting code. Unittests shouldn't plot codes.

File: tftb/tests/test_misc.py [hunk 2]
```diff
@@ -15,6 +15,14 @@
...
 from scipy.signal import argrelmax, argrelmin, hanning
 
+import pylab as plt
+
+# yoder:
+# let's add at least some backwards python2.x compatibility for now.
+import sys
+#py_ver = sys.version_info.major
+# and assume only backwards revisions (for now):
+ispy2 = (sys.version_info.major<3)		# maybe we should work this into TestBase?
```
Comment (jaidevd) [inline tftb/tests/test_misc.py]: Agreed, all these checks should reside in the base classes.

File: tftb/tests/test_misc.py (hunk 1 above)
Comment (jaidevd) [inline tftb/tests/test_misc.py]: Also, this is what seems to be causing the only test failure right now.

File: tftb/generators/analytic_signals.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import numpy as np
 from numpy import pi
 from scipy.signal import hilbert
-from frequency_modulated import fmconst
+from .frequency_modulated import fmconst
```
Comment (jaidevd) [inline tftb/generators/analytic_signals.py]: Ideally we shouldn't use relative imports. Even what I wrote was wrong, this should be:
```python
from tftb.generators.frequency_modulated import fmconst
```

File: tftb/generators/misc.py [hunk 4]
```diff
@@ -56,9 +56,17 @@ def atoms(n_points, coordinates):
...
     # FIXME: This function produces incorrect output when coordinates are
     # one-dimensional.
+    # yoder: would the fix be something like: if not isinstance(coordinates[0], list): coordinates = [coordinates]
+    # ???
+    # at first glance, it looks ok. let's just add some data typ handlers. they may be imperfect, but the basic idea is to catch
+    # some common errors, like passing a list, not an array.
+    # an un-contained 1-d coordinates array:
+    if not hasattr(coordinates[0], '__len__'): coordinates = [coordinates]
+    if not hasattr(coordinates, 'shape'):      coordinates = numpy.array(coordinates)
+    #
```
Comment (jaidevd) [inline tftb/generators/misc.py]: I suggest we fix functions in a separate PR, let's keep this one simply about Python 3 compatibility. What do you think?

File: tftb/generators/misc.py [hunk 5]
```diff
@@ -174,58 +185,72 @@ def mexhat(nu=0.05):
...
+def gdpower(n_points, degree=0.0, rate=1.0):
     """Generate a signal with a power law group delay.
 
     :param n_points: Number of points in time.
     :param degree: degree of the power law.
     :param rate: rate-coefficient of the power law GD.
     :type n_points: int
-    :type degree: int
+    :type degree: int		# yoder: i'm pretty sure this is supposed to be a float. at least in test_misc.py, the test for this funct. passes degree=0.5
     :type rate: float
```
Comment (jaidevd) [inline tftb/generators/misc.py]: You're right. This can be a float too.

File: tftb/generators/misc.py [hunk 6]
```diff
@@ -174,58 +185,72 @@ def mexhat(nu=0.05):
...
 
-    if (degree < 1) and (degree != 0):
-        d = n_points ** (degree * rate)
+    if (degree < 1.) and (degree != 0):
+        #d = n_points ** (degree * rate)
+        d = float(n_points) ** (degree * rate)
         t0 = n_points / 10.0
-        tfx[:lnu] = np.exp(-1j * 2 * pi * (t0 * nu + d * nu ** degree / degree)) * am
+        #tfx[:lnu] = np.exp(-1j * 2 * pi * (t0 * nu + d * nu ** degree / degree)) * am
+        tfx[:lnu] = np.exp(-1.0j * 2. * pi * ((t0 * nu + d * nu ** degree) / float(degree))) * am
```
Comment (jaidevd) [inline tftb/generators/misc.py]: Although this change isn't wrong, it's somewhat unnecessary. `pi` being a float will cast everything else into a float too.

File: tftb/generators/noise.py [hunk 7]
```diff
@@ -24,7 +24,10 @@ def noisecu(n_points):
     if n_points <= 2:
         noise = (np.random.rand(n_points, 1) - 0.5 + 1j * (np.random.rand(n_points, 1) - 0.5)) * np.sqrt(6)
     else:
-        noise = np.random.rand(2 ** nextpow2(n_points),) - 0.5
+        print("debug, n_points: ", n_points)
```
Comment (jaidevd) [inline tftb/generators/noise.py]: Please remove this.

File: tftb/tests/__init__.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+from . import *
+import glob		# file listing utility.
+import os
+#
+__all__ = [os.path.splitext(fl)[0] for fl in glob.glob('*.py')]
```
Comment (jaidevd) [inline tftb/tests/__init__.py]: Why is this file necessary?

File: tftb/tests/base.py [hunk 9]
```diff
@@ -13,9 +13,29 @@
...
+    def __init__(self, *args, **kwargs):
+    	# handle various bits, including some python2-3 compatibility, then execute base __init__ as super()
+    	if not ispy2:
+    		# re-map some function calls:
+    		self.assertItemsEqual = self.assertCountEqual		#(inherited from unittest.TestCase)
+    		# ... and others...
+    		#
+    	# and let's go backwards as well, just to be sure (now, we can correct all the downstream code and remove this hack at a later time...).
+    	if ispy2:
+    		self.assertCountEqual = self.assertItemsEqual
```
Comment (jaidevd) [inline tftb/tests/base.py]: It's not necessary to rewrite the `__init__` method to add aliases to other methods. This should be in the `setUpClass` method of `TestBase`.

File: tftb/tests/test_amplitude_modulations.py [hunk 10]
```diff
@@ -2,6 +2,9 @@
 import numpy as np
 from numpy import pi
 from scipy.signal import argrelmax
+# diagnostics:
+import base		# this would be the 'local' import, so it will depend on the local directory structure. nominally, if we properly configure setup.py
+					# and install, we should be able to use the line below. are tests meant to be part of the package, or run separately?
```
Comment (jaidevd) [inline tftb/tests/test_amplitude_modulations.py]: They're meant to be separate. Simply using nose takes care of this. The following import works with `nosetests`.

File: tftb/utils.py [hunk 11]
```diff
@@ -67,13 +67,20 @@ def izak(x):
...
     return sig
 
 
 def nextpow2(n):
-    """Compute the exponent of the next higher power of 2.
+    """
+    #Compute the exponent of the next higher power of 2.
+    # yoder:
+    # clarifying: returns the next *integer* exponent (as a float) and is lower-inclusive.
+    # ie,:
```
Comment (jaidevd) [inline tftb/utils.py]: Oh, yes. Could you please change the original docstring accordingly?

**Comment by jaidevd** [comment]
@markyoder I like the idea of writing this package such that it's compatible with both Python 2 and 3. To that end, I've added a bunch of comments to your commits. Once they're addressed and the tests have passed, we'll be good to go.

File: tftb/generators/misc.py (hunk 4 above)
Comment (markyoder) [inline tftb/generators/misc.py]: agreed. this can be fixed later, but i recommend leaving some comments to guide that process.

File: tftb/generators/misc.py (hunk 6 above)
Comment (markyoder) [inline tftb/generators/misc.py]: agreed, but i often like to explicitly cast my floats 1) in case something changes in a later revision of the interpreter/compiler, 2) to communicate that the float type is important, and 3) just in case a later revision introduces a grouping that produces a problem, akin to ( int(1)/int(2) )*2.0 ?=  2.0*int(1)/int(2). but that sort of communication can be limited to the code review process as well; your decision.

File: tftb/tests/base.py (hunk 9 above)
Comment (markyoder) [inline tftb/tests/base.py]: ok. i'm not very familiar with the unittest library. either way will work.

**Comment by markyoder** [comment]
ok. i've addressed most of those requests; i think some of them will have to be handled at the discretion of the primary developer -- as per better familiarity with the code base and stylistic considerations. my master branch is updated; i still get a test failure on the test_gdpower() script; maybe i (still) don't have the right array coded in.

**Comment by jaidevd** [comment]
@markyoder Can you give me write access to your branch so I can push some commits and clean this up a bit? I'll skip the gdpower test for now, until we're clear on what's happening.

**Comment by markyoder** [comment]
The standard protocol is to pull down my branch to a local branch on your
system, make changes as needed, merge, then push back to the remote. Or
something like that... Or merge my changes via the pull request, but to a
different branch? If that doesn't work, I'll set up something for you.

On Thursday, January 14, 2016, Jaidev Deshpande <notifications@github.com>
wrote:

> @markyoder <https://github.com/markyoder> Can you give me write access to
> your branch so I can push some commits and clean this up a bit? I'll skip
> the gdpower test for now, until we're clear on what's happening.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-signal/pytftb/pull/82#issuecomment-171872338>.
>


-- 
Mark Yoder, PhD
805 451 8750
https://www.linkedin.com/pub/mark-yoder/11/855/300

"If you bore me, you lose your soul..."
~belly

**Comment by jaidevd** [comment]
That's correct, but the remote that I push to is yours, i.e. `git@github.com:markyoder/pytftb.git`, to which I don't have push access. I could just merge this and then make the changes later, of course. Is that okay?

**Comment by jaidevd** [comment]
I'm only insisting on it because then whatever I do, will be visible on this page, in the context of this PR.

**Comment by markyoder** [comment]
Ok. I'd pull the repo in question (my repo) to a local branch, make
changes, then merge to your local master and push to your remote origin
(aka, your github). But I'll sort out something in the morning.

M

On Thursday, January 14, 2016, Jaidev Deshpande <notifications@github.com>
wrote:

> I'm only insisting on it because then whatever I do, will be visible on
> this page, in the context of this PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-signal/pytftb/pull/82#issuecomment-171873415>.
>


-- 
Mark Yoder, PhD
805 451 8750
https://www.linkedin.com/pub/mark-yoder/11/855/300

"If you bore me, you lose your soul..."
~belly

File: tftb/generators/misc.py (hunk 6 above)
Comment (jaidevd) [inline tftb/generators/misc.py]: Ah! Here's the problem. This right side of this equation is:
```python
np.exp(-1j * 2 * pi * (t0 * nu + d * nu ** degree / degree) * am
```
and what's actually being computed is:
```python
np.exp(-1j * 2 * pi * ((t0 * nu + d * nu ** degree) / degree) * am
```
Also, no need for `float(degree)` since it's been casted in #L203

**Comment by jaidevd** [comment]
Sure, but actually that's not needed anymore. As long as you address the last comment here: https://github.com/scikit-signal/pytftb/pull/82/files#diff-584dc30ef6910a1e6b940929ac847f3aR224

we're good. We'll deal with the rest later.

**Comment by markyoder** [comment]
well then, i think it should be sorted out. note also, as per our
discussion of float() casting: python3 is smarter about floats and integer
division. in python3: 1/2 = .5, whereas in python2.x, 1/2 = 0. also, recall
there is the integer division operator 1//2 = 0.

Mark Yoder, PhD
805 451 8750
https://www.linkedin.com/pub/mark-yoder/11/855/300

"If you bore me, you lose your soul..."
~belly

On Thu, Jan 14, 2016 at 11:30 PM, Jaidev Deshpande <notifications@github.com

**Comment by jaidevd** [comment]
Merging.
